packages/apps/job-launcher/server/src/modules/payment/payment.service.ts

```diff
@@ -120,6 +130,20 @@ export class PaymentService {
       throw new BadRequestException(ErrorPayment.NotSuccess);
     }
 
+    const paymentEntity = await this.paymentRepository.findOne({
```

Comment:
what happen if strip returns FAILED or similar?

Reply:
Payment will stay PENDING in Database.
We can FAILED it also, what do you think about it? 
```public async confirmFiatPayment(
    userId: number,
    data: PaymentFiatConfirmDto,
  ): Promise<boolean> {
    const paymentData = await this.stripe.paymentIntents.retrieve(
      data.paymentId,
    );

    if (!paymentData) {
      this.logger.log(ErrorPayment.NotFound, PaymentService.name);
      throw new NotFoundException(ErrorPayment.NotFound);
    }

    const paymentEntity = await this.paymentRepository.findOne({
      userId,
      transaction: data.paymentId,
      status: PaymentStatus.PENDING
    });

    if (!paymentEntity) {
      this.logger.log(
        ErrorPayment.NotFound,
        PaymentRepository.name,
      );
      throw new BadRequestException(ErrorPayment.NotFound);
    }

    if (paymentData?.status?.toUpperCase() !== PaymentStatus.SUCCEEDED) {
      paymentEntity.status = PaymentStatus.FAILED;
      paymentEntity.save();
      this.logger.log(ErrorPayment.NotSuccess, PaymentService.name);
      throw new BadRequestException(ErrorPayment.NotSuccess);
    }

    const rate = await getRate(Currency.USD, paymentData.currency.toLowerCase());
    
    await this.paymentRepository.create({
      userId,
      source: PaymentSource.FIAT,
      type: PaymentType.DEPOSIT,
      amount: BigNumber.from(paymentData.amount).toString(),
      currency: paymentData.currency.toLowerCase(),
      rate,
      status: PaymentStatus.SUCCEEDED
    })

    return true;
  }```

Reply:
According to [this](https://stripe.com/docs/payments/paymentintents/lifecycle#intent-statuses) the two statuses you can get to set the payment as Failed are:

- requires_payment_method
- canceled

```
public async confirmFiatPayment(
    userId: number,
    data: PaymentFiatConfirmDto,
  ): Promise<boolean> {
    const paymentData = await this.stripe.paymentIntents.retrieve(
      data.paymentId,
    );

    if (!paymentData) {
      this.logger.log(ErrorPayment.NotFound, PaymentService.name);
      throw new NotFoundException(ErrorPayment.NotFound);
    }

    const paymentEntity = await this.paymentRepository.findOne({
      userId,
      transaction: data.paymentId,
      status: PaymentStatus.PENDING
    });

    if (!paymentEntity) {
      this.logger.log(
        ErrorPayment.NotFound,
        PaymentRepository.name,
      );
      throw new BadRequestException(ErrorPayment.NotFound);
    }

    if (paymentData?.status?.toUpperCase() === 'canceled' || 
         paymentData?.status?.toUpperCase() === 'requires_payment_method') {
      paymentEntity.status = PaymentStatus.FAILED;
      paymentEntity.save();
      this.logger.log(ErrorPayment.NotSuccess, PaymentService.name);
      throw new BadRequestException(ErrorPayment.NotSuccess);
    } else if (paymentData?.status?.toUpperCase() !== PaymentStatus.SUCCEEDED) {
      return false; //since it is not being set to FAILED, it can be rechecked later
    }

    const rate = await getRate(Currency.USD, paymentData.currency.toLowerCase());
    
    await this.paymentRepository.create({
      userId,
      source: PaymentSource.FIAT,
      type: PaymentType.DEPOSIT,
      amount: BigNumber.from(paymentData.amount).toString(),
      currency: paymentData.currency.toLowerCase(),
      rate,
      status: PaymentStatus.SUCCEEDED
    })

    return true;
  }
```


Also this should be removed otherwise you would create the payment twice:
```
const rate = await getRate(Currency.USD, paymentData.currency.toLowerCase());

await this.paymentRepository.create({
      userId,
      source: PaymentSource.FIAT,
      type: PaymentType.DEPOSIT,
      amount: BigNumber.from(paymentData.amount).toString(),
      currency: paymentData.currency.toLowerCase(),
      rate,
      status: PaymentStatus.SUCCEEDED
    })
```